File: java/src/main/java/chapter01/Movie.java
package chapter01;

import jdk.nashorn.api.tree.NewTree;

public class Movie {
    public static final int CHILDRENS = 2;
    public static final int REGULAR = 0;
    public static final int NEW_RELEASE = 1;

    private String _title;
    private Price _price;

    public Movie(String title, int priceCode) {
        _title = title;
        setPriceCode(priceCode);
    }

    public int getPriceCode() {
        return _price.getPriceCode();
    }

    public void setPriceCode(int arg) {
        switch (arg) {
            case REGULAR:
                _price = new RegularPrice();
                break;
            case NEW_RELEASE:
                _price = new NewReleasePrice();
                break;
            case CHILDRENS:
                _price = new ChildrensPrice();
                break;
            default:
                throw new IllegalArgumentException("不正な料金コード");
        }
    }

    public String getTitle() {
        return _title;
    }

    double getCharge(int daysRented) {
        double result = 0;
        switch (this.getPriceCode()) {
            case Movie.REGULAR:
                result += 2;
                if (daysRented > 2) {
                    result += (daysRented - 2) * 1.5;
                }
                break;
            case Movie.NEW_RELEASE:
                result += daysRented * 3;
                break;
            case Movie.CHILDRENS:
                result += 1.5;
                if (daysRented > 3) {
                    result += (daysRented - 3) * 1.5;
                }
                break;
        }
        return result;
    }

    int getFrequentRenterPoints(int daysRented) {
        if ((this.getPriceCode() == Movie.NEW_RELEASE) && daysRented > 1)
            return 2;
        else
            return 1;
    }
}

File: go/movie.go
package main

import "fmt"

const CHILDREN = 2
const REGULAR = 0
const NEW_RELEASE = 1

type Movie struct {
	title     string
	priceCode int
}

func NewMovie(title string, priceCode int) Movie {
	return Movie{
		title:     title,
		priceCode: priceCode,
	}
}

func (m Movie) GetPriceCode() int {
	return m.priceCode
}

func (m *Movie) SetPriceCode(priceCode int) {
	m.priceCode = priceCode
}

func (m Movie) GetTitle() string {
	return m.title
}

type Rental struct {
	movie      Movie
	daysRented int
}

func NewRental(movie Movie, daysRented int) Rental {
	return Rental{
		movie:      movie,
		daysRented: daysRented,
	}
}

func (r Rental) GetDaysRented() int {
	return r.daysRented
}

func (r Rental) GetMovie() Movie {
	return r.movie
}

type Customer struct {
	name    string
	rentals []Rental
}

func NewCustomer(name string) Customer {
	return Customer{
		name: name,
	}
}

func (c *Customer) AddRental(r Rental) {
	c.rentals = append(c.rentals, r)
}

func (c Customer) GetName() string {
	return c.name
}

func (c Customer) Statement() string {
	totalAmount := 0.0
	frequentRenterPoints := 0
	rentals := c.rentals
	result := fmt.Sprintf("Rental Record for %s\n", c.GetName())

	for _, v := range rentals {
		thisAmount := v.amountFor()

		frequentRenterPoints++
		if v.GetMovie().GetPriceCode() == NEW_RELEASE &&
			v.GetDaysRented() > 1 {
			frequentRenterPoints++
		}

		result = fmt.Sprintf("%s\t%s\t%.1f\n", result, v.GetMovie().title, thisAmount)
		totalAmount += thisAmount
	}

	result = fmt.Sprintf("%sAmount owed is %.1f\n", result, totalAmount)
	result = fmt.Sprintf("%sYou earned %d frequent renter points\n", result, frequentRenterPoints)
	return result
}

func (r Rental) amountFor() float64 {
	var result float64
	switch r.GetMovie().GetPriceCode() {
	case REGULAR:
		result += 2.0
		if r.GetDaysRented() > 2 {
			result += float64(r.GetDaysRented()-2) * 1.5
		}
	case NEW_RELEASE:
		result += float64(r.GetDaysRented() * 3)
	case CHILDREN:
		result += 1.5
		if r.GetDaysRented() > 3 {
			result += float64(r.GetDaysRented()-3) * 1.5
		}
	}
	return result
}
